### DISCUSSION TITLE
OpenSSL Server Verification Fails on .NET Jammy Chiseled containers

### DISCUSSION THREAD

**Discussion by dlinkov-roblox** [dotnet/dotnet-docker#5371] [Q&A] [answered]
Hello!

We've been pushing for adoption of the chiseled containers, but have run into some difficulties with the profiler we use. We've tracked the issue down to the underlying HTTP library in use [here](https://github.com/yhirose/cpp-httplib). In particular, we are hitting one of the three error cases starting [here](https://github.com/yhirose/cpp-httplib/blob/560854a9611ebe2361bebd7cc60b11a7e76ef2f9/httplib.h#L8781) based on the error logged by the profiler when it tries to export data. The exact same set up works in the equivalent jammy Dockerfile (`dotnet/aspnet:6.0-jammy` vs `dotnet/aspnet:6.0-jammy-chiseled`).

We've tried to copy all of the libraries from a builder stage using the `dotnet/aspnet:6.0-jammy` image. 

```
COPY --from=builder /usr/bin/openssl /usr/bin/openssl
COPY --from=builder /usr/lib/ /usr/lib/
```

This leads us to think this is not a library issue (which had been our hunch). We've also tried to (with no success) change ownership of certain .so files thinking that might be the issue.

Are there any other significant changes between the jammy image and the chiseled jammy image that could lead to this mismatch in behaviour?

Thanks in advance!

**Comment by richlander** [comment]
Can you try running the regular jammy image as non-root? That would help rule out some of the permissions issues.

Do you need a different OpenSSL than the one we include?

**Reply by dlinkov-roblox** [reply]
We've tried using a non-root user on a Jammy container (which worked as expected), and we've also tried running with `USER root` in the chiseled jammy container, which did not work. As you said, that almost definitely rules out permissions issues.

**Reply by richlander** [reply]
Can you tell us more about those two `COPY` lines? They seem suspect to me, in terms of interacting with the underlying image.

Here's the underlying Dockerfile. I'd suggest removing the OpenSSL installation there if you're going to install your own and validating that its dependencies are satisfied. It's easy to build your own customer version.

https://github.com/dotnet/dotnet-docker/blob/main/src/runtime-deps/6.0/jammy-chiseled/amd64/Dockerfile

If that doesn't work, we can jump on a call to try to find a solution.

@cjdcordeiro

**Reply by jaskaransinghdr6j** [reply]
Hey Rich, I work with @dlinkov-roblox. We use `dotnet/aspnet:6.0-jammy-chiseled` for this case. Is there any documentation to how runtime-deps is hooked into the chiseled image so we can try to roll our own? 

Im assuming we need to add `openssl_bins` chisel package.

**Reply by jaskaransinghdr6j** [reply]
We're trying to use this profiler: https://grafana.com/docs/pyroscope/latest/configure-client/language-sdks/dotnet

**Reply by richlander** [reply]
The images are just layered on top of one another.

- https://github.com/dotnet/dotnet-docker/blob/6f32b916ff36e8ff8e3f22b79c0978e4a9392c59/src/aspnet/6.0/jammy-chiseled/amd64/Dockerfile#L17
- https://github.com/dotnet/dotnet-docker/blob/6f32b916ff36e8ff8e3f22b79c0978e4a9392c59/src/runtime/6.0/jammy-chiseled/amd64/Dockerfile#L20
- https://github.com/dotnet/dotnet-docker/blob/6f32b916ff36e8ff8e3f22b79c0978e4a9392c59/src/runtime-deps/6.0/jammy-chiseled/amd64/Dockerfile

You can build these on your own machine with `docker build`. Should just take a few minutes.

I was suggesting that you *remove* OpenSSL since you seem to be copying your own in. Perhaps I misunderstand what those `COPY` lines are doing.

**Reply by jaskaransinghdr6j** [reply]
Gotcha, yes we'll remove openssl and and dependencies we were trying to copy explicitly when trying to chisel this package ourself.

**Reply by cjdcordeiro** [reply]
Hi! As an additional check, just to be sure, can you see if the issue is somehow related to ca-certificates? I.e. the current slices for `ca-certificates` aren't running `update-ca-certificates` (atm) after the slice installation.

**Reply by richlander** [reply]
If you are using Pyroscope, I'm curious why your `COPY` lines don't look more like: https://hub.docker.com/r/pyroscope/pyroscope-dotnet. Is it the case that they didn't work and the two lines you shared are workarounds?

**Reply by jaskaransinghdr6j** [reply, marked as answer]
We were able to resolve this by copying over raw `/usr/share/ca-certificates/` from the builder. We used to run update-ca-certificates in the builder (jammy) and then copy over everything in '/etc/ssl/certs' which includes the `ca-certificate.crt` file. For some reason in this case, the underlying SSL implementation of the httplib also wanted the raw .pem files of the root CAs (we still dont know why)

tldr;

In the jammy builder: 

`RUN update-ca-certificates`

In the runtime chiseled image: 

`COPY --from=builder /usr/share/ca-certificates/ /usr/share/ca-certificates/`
`COPY --from=builder /usr/local/share/ca-certificates/ /usr/local/share/ca-certificates/`
`COPY --from=builder /etc/ssl/certs /etc/ssl/certs`

**Reply by jaskaransinghdr6j** [reply]
This was debugged by looking at STRACE logs to understand which file descriptors are being called in `openat()` syscalls. 

It's still beyond our understanding why raw cets would still be required even after running `update-ca-certificates` and copying the generated concatenated cert to `/etc/ssl/certs`

**Reply by richlander** [reply]
That's a very useful finding. This challenge has come up a few times. Your pattern looks good and that means you can continue using the official images we publish.

We made a similar change for Alpine: https://learn.microsoft.com/dotnet/core/compatibility/containers/8.0/ca-certificates-package

I take it that you are good to go now?

**Reply by jaskaransinghdr6j** [reply]
Thanks for your help @richlander! I think we can close this @dlinkov-roblox

**Comment by dlinkov-roblox** [comment]
Thanks for the help @richlander, this was exceptional support and we really appreciate it!

**Reply by richlander** [reply]
My pleasure. I'm very motivated to help teams adopt chiseled images. We've put a lot of effort into them and think that this form factor is the future. Don't hesitate to reach out for help again.
